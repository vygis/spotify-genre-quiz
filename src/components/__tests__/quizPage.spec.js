import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';

import QuizPage from '../quizPage';

// fix for https://github.com/enzymejs/enzyme/issues/2073
const updateWithTimeout = async (wrapper) => {
  await act(async () => {
    await new Promise(resolve => setTimeout(resolve, 0));
    wrapper.update();
  });
};

describe('quizPage', () => {

  let wrapper;

  describe('when the api errors', () => {

    it('should display the error', async (done) => {
      fetch.mockReject(new Error('Network error'));
      wrapper = mount(<QuizPage/>);
      await(updateWithTimeout(wrapper));
      fetch.mockClear();
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  describe('when the api returns a valid payload', () => {

    beforeEach(async (done) => {
      fetch.mockResponse(JSON.stringify([
        {
          genre: 'rock',
          albumCover: '/sex-drugs-and-sausage-rolls/',
          genreOptions: [{ value: 'rock', isCorrect: true}, { value: 'disco', isCorrect: false }]
        },
        {
          genre: 'opera',
          albumCover: '/pavarotti-lives/',
          genreOptions: [{ value: 'opera', isCorrect: true}, { value: 'techno', isCorrect: false }]
        }
      ]));
      wrapper = mount(<QuizPage/>);
      await updateWithTimeout(wrapper);
      wrapper.find('button').simulate('click');
      fetch.mockClear();
      done();
    });

    it('should display the game when starting', () => {
      expect(wrapper.find('Game')).toMatchSnapshot();
    });

    it('should indicate when a question is answered correctly', () => {
      wrapper.find({ answer: { isCorrect: true }}).simulate('click');
      expect(wrapper.find('AnswerResult')).toMatchSnapshot();
    });

    it('should indicate when a question is answered incorrectly', () => {
      wrapper.find({ answer: { isCorrect: false }}).simulate('click');
      expect(wrapper.find('AnswerResult')).toMatchSnapshot();
    });

    it('should display the game result when all questions are answered, and restart the game', async (done) => {
      wrapper.find({ answer: { isCorrect: false }}).simulate('click');
      wrapper.find('button').simulate('click');
      wrapper.find({ answer: { isCorrect: true }}).simulate('click');
      wrapper.find('button').simulate('click');
      expect(wrapper.find('GameResult')).toMatchSnapshot();

      fetch.mockResponse(JSON.stringify([
        {
          genre: 'country',
          albumCover: '/been-cleaning-the-barn/',
          genreOptions: [{ value: 'country', isCorrect: true}, { value: 'pop', isCorrect: false }]
        }
      ]));
      wrapper.find('button').simulate('click');
      await(updateWithTimeout(wrapper));
      expect(wrapper.find('Game')).toMatchSnapshot();
      done();
    });
  });
});
