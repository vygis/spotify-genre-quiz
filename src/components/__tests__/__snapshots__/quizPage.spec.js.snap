// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quizPage when the api errors should display the error 1`] = `
<QuizPage>
  <div
    className="col-12 pt-3"
  >
    <span
      className="alert alert-danger"
    >
      Network error
      . Try reloading the page.
    </span>
  </div>
</QuizPage>
`;

exports[`quizPage when the api returns a valid payload should display the game result when all questions are answered, and restart the game 1`] = `
<GameResult
  answers={
    Array [
      Object {
        "albumCover": "/sex-drugs-and-sausage-rolls/",
        "genre": "rock",
        "genreOptions": Array [
          Object {
            "isCorrect": true,
            "value": "rock",
          },
          Object {
            "isCorrect": false,
            "value": "disco",
          },
        ],
        "isAnswerCorrect": false,
      },
      Object {
        "albumCover": "/pavarotti-lives/",
        "genre": "opera",
        "genreOptions": Array [
          Object {
            "isCorrect": true,
            "value": "opera",
          },
          Object {
            "isCorrect": false,
            "value": "techno",
          },
        ],
        "isAnswerCorrect": true,
      },
    ]
  }
  isLoading={false}
  onClickFn={[Function]}
>
  <div>
    <h2>
      Quiz completed!
    </h2>
    <p>
      You answered 
      1
       out of 
      2
       questions correctly.
    </p>
    <div>
      <LoadingButton
        isLoading={false}
        onClickFn={[Function]}
        title="Play again"
      >
        <button
          className="btn btn-dark rounded-pill loading-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span
            style={
              Object {
                "visibility": "unset",
              }
            }
          >
            Play again
          </span>
        </button>
      </LoadingButton>
    </div>
  </div>
</GameResult>
`;

exports[`quizPage when the api returns a valid payload should display the game result when all questions are answered, and restart the game 2`] = `
<Game
  isComplete={false}
  isLoading={false}
  onCompleteFn={[Function]}
  onRestartFn={[Function]}
  questionCount={1}
  questions={
    Array [
      Object {
        "albumCover": "/been-cleaning-the-barn/",
        "genre": "country",
        "genreOptions": Array [
          Object {
            "isCorrect": true,
            "value": "country",
          },
          Object {
            "isCorrect": false,
            "value": "pop",
          },
        ],
      },
    ]
  }
>
  <section>
    <div
      className="card mx-auto"
    >
      <Question
        isLastQuestion={true}
        onAnswerSelect={[Function]}
        onNext={[Function]}
        question={
          Object {
            "albumCover": "/been-cleaning-the-barn/",
            "genre": "country",
            "genreOptions": Array [
              Object {
                "isCorrect": true,
                "value": "country",
              },
              Object {
                "isCorrect": false,
                "value": "pop",
              },
            ],
          }
        }
      >
        <div>
          <div
            className="card-img-top card-img-wrapper"
          >
            <img
              src="data:image/jpeg;base64,/been-cleaning-the-barn/"
            />
          </div>
          <div
            className="card-body"
          >
            <div
              className="card-text"
            >
              <AnswerOption
                answer={
                  Object {
                    "isCorrect": true,
                    "value": "country",
                  }
                }
                discloseResult={false}
                key="0"
                onClickFn={[Function]}
              >
                <span
                  className="badge badge-pill badge-dark text-uppercase "
                  onClick={[Function]}
                >
                  country
                </span>
              </AnswerOption>
              <AnswerOption
                answer={
                  Object {
                    "isCorrect": false,
                    "value": "pop",
                  }
                }
                discloseResult={false}
                key="1"
                onClickFn={[Function]}
              >
                <span
                  className="badge badge-pill badge-dark text-uppercase "
                  onClick={[Function]}
                >
                  pop
                </span>
              </AnswerOption>
            </div>
          </div>
        </div>
      </Question>
      <GameProgress
        total={1}
        upcoming={0}
      >
        <div
          className="card-footer pt-3"
        >
          <div
            className="card-footer progress-bar"
            style={
              Object {
                "width": 0,
              }
            }
          />
        </div>
      </GameProgress>
    </div>
  </section>
</Game>
`;

exports[`quizPage when the api returns a valid payload should display the game when starting 1`] = `
<Game
  isComplete={false}
  isLoading={false}
  onCompleteFn={[Function]}
  onRestartFn={[Function]}
  questionCount={2}
  questions={
    Array [
      Object {
        "albumCover": "/sex-drugs-and-sausage-rolls/",
        "genre": "rock",
        "genreOptions": Array [
          Object {
            "isCorrect": true,
            "value": "rock",
          },
          Object {
            "isCorrect": false,
            "value": "disco",
          },
        ],
      },
      Object {
        "albumCover": "/pavarotti-lives/",
        "genre": "opera",
        "genreOptions": Array [
          Object {
            "isCorrect": true,
            "value": "opera",
          },
          Object {
            "isCorrect": false,
            "value": "techno",
          },
        ],
      },
    ]
  }
>
  <section>
    <div
      className="card mx-auto"
    >
      <Question
        isLastQuestion={false}
        onAnswerSelect={[Function]}
        onNext={[Function]}
        question={
          Object {
            "albumCover": "/sex-drugs-and-sausage-rolls/",
            "genre": "rock",
            "genreOptions": Array [
              Object {
                "isCorrect": true,
                "value": "rock",
              },
              Object {
                "isCorrect": false,
                "value": "disco",
              },
            ],
          }
        }
      >
        <div>
          <div
            className="card-img-top card-img-wrapper"
          >
            <img
              src="data:image/jpeg;base64,/sex-drugs-and-sausage-rolls/"
            />
          </div>
          <div
            className="card-body"
          >
            <div
              className="card-text"
            >
              <AnswerOption
                answer={
                  Object {
                    "isCorrect": true,
                    "value": "rock",
                  }
                }
                discloseResult={false}
                key="0"
                onClickFn={[Function]}
              >
                <span
                  className="badge badge-pill badge-dark text-uppercase "
                  onClick={[Function]}
                >
                  rock
                </span>
              </AnswerOption>
              <AnswerOption
                answer={
                  Object {
                    "isCorrect": false,
                    "value": "disco",
                  }
                }
                discloseResult={false}
                key="1"
                onClickFn={[Function]}
              >
                <span
                  className="badge badge-pill badge-dark text-uppercase "
                  onClick={[Function]}
                >
                  disco
                </span>
              </AnswerOption>
            </div>
          </div>
        </div>
      </Question>
      <GameProgress
        total={2}
        upcoming={0}
      >
        <div
          className="card-footer pt-3"
        >
          <div
            className="card-footer progress-bar"
            style={
              Object {
                "width": 0,
              }
            }
          />
        </div>
      </GameProgress>
    </div>
  </section>
</Game>
`;

exports[`quizPage when the api returns a valid payload should indicate when a question is answered correctly 1`] = `
<AnswerResult
  isAnswerCorrect={true}
  isLastQuestion={false}
  onClickFn={[Function]}
>
  <div
    className="card-overlay"
  >
    <div
      className="display-4"
    >
      CORRECT!
    </div>
    <button
      className="btn btn-outline-light rounded-pill"
      onClick={[Function]}
      type="button"
    >
      Next
    </button>
  </div>
</AnswerResult>
`;

exports[`quizPage when the api returns a valid payload should indicate when a question is answered incorrectly 1`] = `
<AnswerResult
  isAnswerCorrect={false}
  isLastQuestion={false}
  onClickFn={[Function]}
>
  <div
    className="card-overlay"
  >
    <div
      className="display-4"
    >
      WRONG!
    </div>
    <button
      className="btn btn-outline-light rounded-pill"
      onClick={[Function]}
      type="button"
    >
      Next
    </button>
  </div>
</AnswerResult>
`;
